src/graph/sampling/metapath_randomwalk.h

```diff
@@ -0,0 +1,147 @@
+using namespace dgl::aten;
+
+namespace sampling {
+
+namespace impl {
+
+namespace {
+
+// bool WhetherToTerminate(
+//     void *node_ids_generated_so_far,
+//     dgl_id_t last_node_id_generated,
+//     int64_t number_of_nodes_generated_so_far)
+using TerminatePredicate = std::function<bool(void *, dgl_id_t, int64_t)>;
+
+/*!
+ * \brief Select one successor of metapath-based random walk, given the path generated
+ * so far.
+ *
+ * \param data The path generated so far, of type \c IdxType.
+ * \param curr The last node ID generated.
+ * \param len The number of nodes generated so far.  Note that the seed node is always
+ *            included as \c data[0], and the successors start from \c data[1].
+ *
+ * \param edges_by_type Vector of results from \c GetAdj() by edge type.
+ * \param metapath_data Edge types of given metapath.
+ * \param prob Transition probability per edge type.
+ * \param terminate Predicate for terminating the current random walk path.
+ *
+ * \return A pair of ID of next successor (-1 if not exist), as well as whether to terminate.
+ */
+template<DLDeviceType XPU, typename IdxType>
+std::pair<dgl_id_t, bool> MetapathRandomWalkStep(
+    void *data,     // void * because std::function does not accept template typenames.
+    dgl_id_t curr,
+    int64_t len,
+    const std::vector<std::vector<IdArray> > &edges_by_type,
+    const IdxType *metapath_data,
+    const std::vector<FloatArray> &prob,
+    TerminatePredicate terminate) {
+  dgl_type_t etype = metapath_data[len];
+
+  // Note that since the selection of successors is very lightweight (especially in the
+  // uniform case), we want to reduce the overheads (even from object copies or object
+  // construction) as much as possible.
+  // Using Successors() slows down by 2x.
+  // Using OutEdges() slows down by 10x.
+  const auto &csr_arrays = edges_by_type[etype];
+  const IdxType *offsets = static_cast<IdxType *>(csr_arrays[0]->data);
+  const IdxType *all_succ = static_cast<IdxType *>(csr_arrays[1]->data);
+  const IdxType *succ = all_succ + offsets[curr];
+
+  int64_t size = offsets[curr + 1] - offsets[curr];
+  if (size == 0)
+    return std::make_pair(-1, true);
+
+  FloatArray prob_etype = prob[etype];
+  IdxType idx;
+  if (prob_etype->shape[0] == 0) {
+    // empty probability array; assume uniform
+    idx = RandomEngine::ThreadLocal()->RandInt(size);
+  } else {
+    // non-uniform random walk
+    const IdxType *all_eids = static_cast<IdxType *>(csr_arrays[2]->data);
+    const IdxType *eids = all_eids + offsets[curr];
+
+    ATEN_FLOAT_TYPE_SWITCH(prob_etype->dtype, DType, "probability", {
+      FloatArray prob_selected = FloatArray::Empty({size}, prob_etype->dtype, prob_etype->ctx);
+      DType *prob_selected_data = static_cast<DType *>(prob_selected->data);
+      const DType *prob_etype_data = static_cast<DType *>(prob_etype->data);
+      for (int64_t j = 0; j < size; ++j)
+        prob_selected_data[j] = prob_etype_data[eids[j]];
```

Comment:
Is this an `IndexSelect` operation?

Reply:
Using `CreateView` and `IndexSelect` slows it down by 30%.

Reply:
Here is what I tried to replace the code above with:
```C++
    // XXX CreateView() mutates the (refcount of the) original array but here we have a const.
    const IdArray eids = const_cast<IdArray &>(csr_arrays[2]).CreateView(
        {offsets[curr + 1] - offsets[curr]},
        csr_arrays[2]->dtype,
        offsets[curr] * sizeof(IdxType));
    const FloatArray prob_selected = std::move(IndexSelect(prob_etype, eids));
    idx = RandomEngine::ThreadLocal()->Choice<IdxType>(prob_selected);
```

Reply:
I c. Let's use your solution. BTW, the const cast could be avoided by copying out `csr_arrays[2]`, which is internally a shallow copy.